src/non_expressive_names.rs

```diff
@@ -0,0 +1,218 @@
+use syntax::visit::{self, FnKind};
+use utils::{span_lint_and_then, in_macro, span_lint};
+use strsim::levenshtein;
+
+/// **What it does:** This lint warns about names that are very similar and thus confusing
+///
+/// **Why is this bad?** It's hard to distinguish between names that differ only by a single character
+///
+/// **Known problems:** None?
+///
+/// **Example:** `checked_exp` and `checked_expr`
+declare_lint! {
+    pub SIMILAR_NAMES,
+    Warn,
+    "similarly named items and bindings"
+}
+
+/// **What it does:** This lint warns about having too many variables whose name consists of a single character
+///
+/// **Why is this bad?** It's hard to memorize what a variable means without a descriptive name.
+///
+/// **Known problems:** None?
+///
+/// **Example:** let (a, b, c, d, e, f, g) = (...);
+declare_lint! {
+    pub MANY_SINGLE_CHAR_NAMES,
+    Warn,
+    "too many single character bindings"
+}
+
+pub struct NonExpressiveNames {
+    pub similarity_threshold: usize,
+    pub max_single_char_names: usize,
+}
+
+impl LintPass for NonExpressiveNames {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(SIMILAR_NAMES, MANY_SINGLE_CHAR_NAMES)
+    }
+}
+
+struct SimilarNamesLocalVisitor<'a, 'b: 'a> {
+    names: Vec<(InternedString, Span)>,
+    cx: &'a EarlyContext<'b>,
+    lint: &'a NonExpressiveNames,
+    single_char_names: Vec<char>,
+}
+
+const WHITELIST: &'static [&'static str] = &[
+    "lhs", "rhs",
+];
+
+struct SimilarNamesNameVisitor<'a, 'b: 'a, 'c: 'b>(&'a mut SimilarNamesLocalVisitor<'b, 'c>);
+
+impl<'v, 'a, 'b, 'c> visit::Visitor<'v> for SimilarNamesNameVisitor<'a, 'b, 'c> {
+    fn visit_pat(&mut self, pat: &'v Pat) {
+        if let PatKind::Ident(_, id, _) = pat.node {
+            self.check_name(id.span, id.node.name);
+        }
+        visit::walk_pat(self, pat);
+    }
+}
+
+impl<'a, 'b, 'c> SimilarNamesNameVisitor<'a, 'b, 'c> {
+    fn check_name(&mut self, span: Span, name: Name) {
+        if in_macro(self.0.cx, span) {
+            return;
+        }
+        let interned_name = name.as_str();
+        if interned_name.chars().any(char::is_uppercase) {
+            return;
+        }
+        let count = interned_name.chars().count();
+        if count < 3 {
+            if count != 1 {
+                return;
+            }
+            let c = interned_name.chars().next().expect("already checked");
+            // make sure we ignore shadowing
+            if self.0.single_char_names.contains(&c) {
+                return;
+            }
+            self.0.single_char_names.push(c);
+            if self.0.single_char_names.len() < self.0.lint.max_single_char_names {
+                return;
+            }
+            span_lint(self.0.cx,
+                      MANY_SINGLE_CHAR_NAMES,
+                      span,
+                      &format!("{}th binding whose name is just one char",
+                               self.0.single_char_names.len()));
+            return;
```

Comment:
yea, that should be `if foo { a } else { b } return;` iirc we have a lint that detects similar cases, I'll see if I can improve it.